doc/docusaurus/docs/essential-concepts/versions.md

```diff
@@ -0,0 +1,68 @@
+---
+sidebar_position: 5
+---
+
+# Different Notions of Version
+
+There are several different notions of version that Cardano smart contract developers must distinguish.
+
+## Ledger Language Version
+
+This is what "Plutus V1", "Plutus V2", "Plutus V3" refer to.
+It is called "ledger" language version because they are _not_ really different versions of a language in the sense of programming languages, but different languages from the Cardano ledger's point of view; the ledger handles Plutus V1 vs. V2 vs. V3 differently.
+In essence, "Plutus V1", "Plutus V2" and "Plutus V3" are tags that the ledger attaches to validators in a transaction.
+
+From the ledger's perspective, they are in fact totally different languages:
+there is no requirement that they be similar or compatible in any way.
+However, the "V1", "V2" and "V3" naming scheme is practically useful because in reality they have a lot in common (e.g., the underlying language is Plutus Core and programs are executed by the Plutus Core evaluator, regardless of ledger language version); plus it makes it clear the order that they were introduced and the relationships among them.
+
+When simply writing a standalone program, say one that takes two integers and returns their sum, or one that implements a Sudoku solver, the notion of ledger language version is completely irrelevant.
+There is no need (and nowhere) to specify whether you are writing a Plutus V1, or Plutus V2, or Plutus V3 program - again, the concept of ledger language version does not apply to programs themselves.
```

Comment:
As I said in my other comment - in the near future all other differences between V1, V2 and V3 will go away, and the only difference left would be: they receive different arguments from the ledger. There is no fundamental reason for there to be different programming language features among V1/V2/V3.

Reply:
> As I said in my other comment - in the near future all other differences between V1, V2 and V3 will go away, and the only difference left would be: they receive different arguments from the ledger. 

How can this possibly be true? We _compute_ the semantics variant from the ledger language version (and the protocol version), it cannot be true that the only difference is in arguments, since the semantics variant affects the behavior of programs. The semantics variant is a derivative and in my view is a vestige of the past. It's the ledger language version and the protocol version that are real, the semantics variant is just a made up hash of the two.

In particular, V1 (or V2) and V3 _behave differently_ for certain programs with a _fixed_ protocol version. How is that not a difference? Sure we arrange that through the semantics variant, but that's just a detail of the implementation and not at all how we actually look at versioning ("we'll change this behavior in V3").

Reply:
The Plutus Core language can exist as a standalone programming language, without the node, ledger, or anything Cardano. It is turing complete and you can use it to do anything. For that reason, builtin semantics variant should be the first-class concept. When we talk about the language without Cardano, there is no need to introduce the concept of PlutusV1, V2, V3 etc.

The overwhelmingly main reason to introduce a new PlutusVx is due to a new ledger era, necessitating changes to the script context or other script arguments or return types. Admittedly, there's another potential reason to introduce a new PlutusVx, without a new ledger era, namely changing the behavior of a builtin function. But crucially, the only reason a new PlutusVx is needed in this case, rather than just changing the behavior at the next Protocol Version, _is still Cardano-related_: because of the nature of the blockchain, we want scripts that are already submitted to continue to have the old behavior, thus new PlutusVx, rather than new Protocol Version.

Without Cardano in the picture, not only would we not need Plutus V1/V2/V3, but we certainly wouldn't name them Plutus V1/V2/V3. How is it a better name than Builtin Semantics Variant V1/V2/V3? The latter is much more specific and accurate, and less confusing.

Reply:
btw one more point: the fact that the ledger language version and protocol version must be used together to determine the behaviors of builtin functions is another reason why the ledger language version is irrelevant when we are talking about the language itself, independent of Cardano: protocol versions only exist in Cardano, and without protocol version, the ledger language version does not even determine anything by itself.

Reply:
I agree with everything you've said apart from

> Without Cardano in the picture, not only would we not need Plutus V1/V2/V3, but we certainly wouldn't name them Plutus V1/V2/V3. How is it a better name than Builtin Semantics Variant V1/V2/V3? The latter is much more specific and accurate, and less confusing.

(as I'm arguing in those threads that I've referenced earlier, making the semantics variant in Cardano a number/letter is a bad idea that we or our successors will very likely backtrack on. We think of it as a product type and it really should be a product).

I just do not find your point of view helpful for users, i.e. those people for whom these docs are written for. We're talking about Cardano here, why should we generalize anything beyond that?

> without protocol version, the ledger language version does not even determine anything by itself.

Technically,

1. `PlutusV3` solely determines the semantics variant
2. for earlier ledger language versions they determine the _set_ of available semantics variant

But I see your point.

> For that reason, builtin semantics variant should be the first-class concept. 

I agree. I don't think it's helpful for these docs though. In Cardano the semantics variant is a combination of the ledger language version and the protocol version, nothing more than that. Hashing them into that `A`/`B`/`C` thing is just counter-productive and so is saying things like

> When simply writing a standalone program, say one that takes two integers and returns their sum, or one that implements a Sudoku solver, the notion of ledger language version is completely irrelevant.

Replace `addInteger` in that example with `consByteString` (or even `multiplyInteger`!) and it simply falls apart. You're presuming that there's an `addInteger` operation, a very much Cardano-specific one, and then you're saying that the ledger language is completely irrelevant, which is simply not true, because it is relevant for the behavior of builtins. There's no Sudoku solver if you don't know how your builtins are supposed to behave, it's all one giant undefined behavior agent in that case.

Reply:
> I just do not find your point of view helpful for users, i.e. those people for whom these docs are written for. We're talking about Cardano here, why should we generalize anything beyond that?

I find it easier to think about this in terms of two layers: (1) what is the language itself and the notions related only to the language; and then (2) how it works with Cardano, which is in principle just one of the applications of Plutus Core. Mixing them together makes it harder to untangle the different concepts.

> Replace addInteger in that example with consByteString (or even multiplyInteger!) and it simply falls apart.

With `consByteString`, the ledger language version is still irrelevant in the sense that... it simply isn't a thing when talking about the language itself (layer 1). But I see your point: when running the script on Cardano, different ledger language versions do lead to different behaviors. My argument still holds but I could've phrase it more accurately.

Reply:
> I find it easier to think about this in terms of two layers: (1) what is the language itself and the notions related only to the language; and then (2) how it works with Cardano, which is in principle just one of the applications of Plutus Core.

I agree that thinking in terms of two layers is best for _us_. I really disagree that it's helpful to users. In particular, ironically

> Mixing them together makes it harder to untangle the different concepts.

mixing things up is more likely when you try to untangle them. This shows w.r.t. my point about `consByteString`, but also more generally when you say things like "when simply writing a standalone program". There's no such thing as a program written in some abstract Plutus, it's always some concrete form of Plutus and in this case it's Cardano Plutus (I'm not even sure if other Plutuses exist). Plutus-the-library is a language constructor: you can have your own builtins, your own versioning for them, your own versioning for the AST etc. But the moment you start "writing a program", these become fixed and you can no longer untangle it from its context, i.e. Cardano in our case.

This also shows in

> You simply write your code in Plutus Tx, compile it to UPLC, and run it with a UPLC evaluator.

which, as I said earlier, is merely a bug in the implementation.

So no, I do not agree with

> My argument still holds

and I most certainly do not agree with

> I could've phrase it more accurately.

because these are supposed to be user-friendly docs, not the kind of docs that we write for ourselves with all the gory details and whatnot. All that phrased correctly context is going to be lost on folks who just want to know how to develop smart contracts. E.g. they will just think that the behavior of Plinth scripts is always the same regardless of the ledger language version, which is very much not the case.

Summarizing:

1. I believe that the ledger language version (likely it's "condensed" form) should be an explicit (rather than implicit) part of the semantics variant
2. Plinth users, i.e. Cardano users, should be able to control the ledger language version that they're using for their Plinth scripts right in those scripts
3. while the two-layered model that we've been discussing here is correct, it's counter-productive to put it in these docs as it adds a lot of unnecessary confusion. I really don't feel like this level of detail belongs to an "essential concepts" section

I think one way forward would be to ask the smart contract developers community to read the docs and tell us whether they find them helpful and what their takeaways are.

Reply:
>  I really disagree that it's helpful to users

I guess we'll agree to disagree. I definitely find discussing different things separately useful, whether I'm the developer or the user. Not dissimilar to modularity of code. Plus, this is something we are already doing: the documentation has a section talking about "using Plutus Tx", i.e., just language stuff, and another section for "working with scripts", i.e., Cardano-related. I think it's much better than merging them together - I fail to see how the latter improves readability.

> All that phrased correctly context is going to be lost on folks who just want to know how to develop smart contracts

It's IMO fairly straightforward: when you are writing a standlone program, you can use builtin semantics variant to determine the builtin behavior. When you are writing a script, this is determined by (protocol version, ledger language version). Not that hard!